### PR TITLE
[JENKINS-49994] Infrastructure for warning about serialization of anonymous inner classes

### PR DESCRIPTION
**Pull request by jglick**
[JENKINS-49994](https://issues.jenkins-ci.org/browse/JENKINS-49994)

@reviewbybees

### CODE REVIEW

File: src/main/java/org/jenkinsci/remoting/util/AnonymousClassWarnings.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+                return; // fast path
+            }
+        }
+        Channel channel = Channel.current();
+        if (channel == null) {
+            doCheck(clazz);
+        } else {
+            // May not call methods like Class#isAnonymousClass synchronously, since these can in turn trigger remote class loading.
+            try {
+                channel.executor.submit((Runnable) () -> doCheck(clazz));
```
Comment (jglick) [inline src/main/java/org/jenkinsci/remoting/util/AnonymousClassWarnings.java]: Having this `ExecutorService` be passed to `check` would be cleaner, but I did not see a reasonable way to do that—a ton of other APIs would have to be modified to thread this parameter in. The use of `Channel.current()` suffices for our purposes.

File: src/main/java/org/jenkinsci/remoting/util/AnonymousClassWarnings.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        String name = c.getName();
+        String codeSource = codeSource(c);
+        // Need to be very defensive about calling anything while holding this lock, lest we trigger class loading-related deadlocks.
+        boolean doWarn;
+        synchronized (checked) {
+            doWarn = checked.put(c, true) == null;
+        }
+        if (doWarn) {
+            if (codeSource == null) {
+                // TODO create jenkins.io/redirect
```
Comment (jglick) [inline src/main/java/org/jenkinsci/remoting/util/AnonymousClassWarnings.java]: https://github.com/jenkins-infra/jenkins.io/pull/1439

File: src/main/java/org/jenkinsci/remoting/util/AnonymousClassWarnings.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        boolean doWarn;
+        synchronized (checked) {
+            doWarn = checked.put(c, true) == null;
+        }
+        if (doWarn) {
+            if (codeSource == null) {
+                LOGGER.warning("Attempt to (de-)serialize " + kind + " class " + name + "; see: https://jenkins.io/redirect/serialization-of-anonymous-classes/");
+            } else {
+                // most easily tracked back to source using javap -classpath <location> -l '<name>'
+                LOGGER.warning("Attempt to (de-)serialize " + kind + " class " + name + " in " + codeSource + "; see: https://jenkins.io/redirect/serialization-of-anonymous-classes/");
```
Comment (jglick) [inline src/main/java/org/jenkinsci/remoting/util/AnonymousClassWarnings.java]: example format:

```
Attempt to (de-)serialize anonymous class hudson.remoting.ClassFilterTest$3 in file:/…/remoting/target/test-classes/; see: https://jenkins.io/redirect/serialization-of-anonymous-classes/
```

**Comment by rysteboe** [comment]
Err, missed the CI failures.  Investigating.

**Comment by rysteboe** [comment]
Tests pass, the build failures looks like it's unrelated.

**Pull request merged**